### PR TITLE
MM-32351: Add Compliance Subsections

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
Adding the sub sections for Compliance section in system console

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-32351

#### Release Note
```release-note
Add Compliance Subsections and Deprecate Compliance section Permission
```

### CODE REVIEW

File: model/permission.go [hunk 1]
```diff
@@ -888,28 +938,66 @@ func initializePermissions() {
...
-		"sysconsole_read_plugins",
-		"authentication.permissions.use_group_mentions.name",
-		"authentication.permissions.use_group_mentions.description",
+	PERMISSION_SYSCONSOLE_READ_COMPLIANCE_DATA_RETENTION = &Permission{
+		"sysconsole_read_compliance_data_retention",
+		"",
+		"",
 		PermissionScopeSystem,
 	}
-	PERMISSION_SYSCONSOLE_WRITE_PLUGINS = &Permission{
```
Comment (BenCookie95) [inline model/permission.go:909]: Did you mean to remove the read and write for plugins?

**Review by BenCookie95 — CHANGES_REQUESTED** [review]
Looks good, just some failing tests to fix and one comment!

**Comment by BenCookie95** [comment]
Gave this a quick test locally. I gave a user edit access to data retention and read only to the other compliance sections.

My user could read all the compliance sections, edit the data retention policy but was unable to run the data retention job. The job endpoint was giving me this response:
```
{
     detailed_error: "userId=ofikddfst3remn5sphfstg1o3c, permission=create_data_retention_job,",
     id: "api.context.permissions.app_error",
     message: "You do not have the appropriate permissions.",
     request_id: "mg8ktzu3p78hfk34gwkgzwdp1w",
     status_code: 403,
}
```

File: model/role.go [hunk 2]
```diff
@@ -115,6 +108,23 @@ func init() {
 		PERMISSION_SYSCONSOLE_WRITE_SITE.Id: {
 			PERMISSION_EDIT_BRAND,
 		},
+		PERMISSION_SYSCONSOLE_WRITE_COMPLIANCE_DATA_RETENTION.Id: {
+			PERMISSION_CREATE_DATA_RETENTION_JOB,
+		},
+		PERMISSION_SYSCONSOLE_READ_COMPLIANCE_DATA_RETENTION.Id: {
+			PERMISSION_READ_JOBS,
```
Comment (maxerenberg) [inline model/role.go:115]: Shouldn't this be PERMISSION_READ_DATA_RETENTION_JOB and PERMISSION_READ_COMPLIANCE_EXPORT_JOB?

File: model/role.go [hunk 3]
```diff
@@ -115,6 +108,23 @@ func init() {
...
+		},
+		PERMISSION_SYSCONSOLE_WRITE_COMPLIANCE_COMPLIANCE_EXPORT.Id: {
+			PERMISSION_CREATE_COMPLIANCE_EXPORT_JOB,
+			PERMISSION_DOWNLOAD_COMPLIANCE_EXPORT_RESULT,
+		},
+		PERMISSION_SYSCONSOLE_READ_COMPLIANCE_COMPLIANCE_EXPORT.Id: {
+			PERMISSION_READ_JOBS,
```
Comment (maxerenberg) [inline model/role.go:122]: And likewise, PERMISSION_READ_COMPLIANCE_EXPORT_JOB?

File: app/job.go [hunk 4]
```diff
@@ -80,7 +82,32 @@ func (a *App) SessionHasPermissionToCreateJob(session model.Session, job *model.
 		model.JOB_TYPE_EXPORT_DELETE,
 		model.JOB_TYPE_CLOUD:
 		return a.SessionHasPermissionTo(session, model.PERMISSION_MANAGE_JOBS), model.PERMISSION_MANAGE_JOBS
+	}
 
+	return false, nil
+}
+
+func (a *App) SessionHasPermissionToReadJob(session model.Session, job *model.Job) (bool, *model.Permission) {
```
Comment (maxerenberg) [inline app/job.go:90]: Where is this function used?

File: model/role.go (hunk 2 above)
Comment (hahmadia) [reply model/role.go:115]: Nice catch @maxerenberg

File: model/role.go (hunk 3 above)
Comment (hahmadia) [reply model/role.go:122]: Nice catch @maxerenberg

File: model/permission.go (hunk 1 above)
Comment (sbishel) [reply model/permission.go:909]: Yes. It was in there twice. Also on line 917.

**Comment by hahmadia** [comment]
Hi @sbishel 
Before reviewing this PR, would you be kind enough to fix the breaking tests in the CI pipeline. Thank you!

**Review by hahmadia — COMMENTED** [review]
Hi @sbishel
Before reviewing this PR, would you be able to fix the breaking tests in the CI pipeline. Thank you!

File: api4/job.go [hunk 5]
```diff
@@ -134,12 +140,19 @@ func getJobs(c *Context, w http.ResponseWriter, r *http.Request) {
...
+			mlog.Warn("The job types of a job you are trying to retrieve does not contain permissions", mlog.String("jobType", jobType))
+			continue
+		}
+		if hasPermission {
+			validJobTypes = append(validJobTypes, jobType)
+		}
 	}
 
-	jobs, err := c.App.GetJobsPage(c.Params.Page, c.Params.PerPage)
+	jobs, err := c.App.GetJobsByTypesPage(validJobTypes, c.Params.Page, c.Params.PerPage)
```
Comment (sbishel) [inline api4/job.go:155]: There were two ways I could have implemented this. 
1. Determine JobTypes session has access to, then query for those JobTypes.
- Good - (Performance) Allows the database to handle filtering (there IS an index on type)
- Bad - (Maintenance) Requires maintaining model.ALL_JOB_TYPES.
2. Return all Jobs and Loop
- Good - (Maintenance) Less changes, will work automatically with new JobTypes
- Bad - (Performance) Although done in paging, all Jobs are returned therefore could be looping through thousands of jobs.

Additional Information - The GetJobs() endpoint is not used by our Web App.

File: app/job.go (hunk 4 above)
Comment (sbishel) [reply app/job.go]: Thanks, I lost the changes for this. Not sure what happened but added now.

**Comment by sbishel** [comment]
> Gave this a quick test locally. I gave a user edit access to data retention and read only to the other compliance sections.
> 
> My user could read all the compliance sections, edit the data retention policy but was unable to run the data retention job. The job endpoint was giving me this response:
> 
> ```
> {
>      detailed_error: "userId=ofikddfst3remn5sphfstg1o3c, permission=create_data_retention_job,",
>      id: "api.context.permissions.app_error",
>      message: "You do not have the appropriate permissions.",
>      request_id: "mg8ktzu3p78hfk34gwkgzwdp1w",
>      status_code: 403,
> }
> ```

This will require the changes Hossein made in this PR. 
https://github.com/mattermost/mattermost-server/pull/17061

**Review by hahmadia — APPROVED** [review]
Looks great Scott! Thank you ❤️

**Review by maxerenberg — APPROVED** [review]

**Review by BenCookie95 — APPROVED** [review]

**Comment by furqanmlk** [comment]
Hi @sbishel 

Can you please resolve conflict ?

**Comment by hahmadia** [comment]
Tested as a System Admin and System Manager. 
System manager was No Access, Read Only, Full Access. 
When I had each of those permissions, I got expected behaviour (Not being able to call end point, not seeing the side bar).
Thanks Scott.

**Pull request merged**